### ISSUE TITLE
`AsName.name` has incorrect type annotations

### ISSUE THREAD

**Issue by zsol** [Instagram/LibCST#503]
For example it can be a `Subscript`, not just `Name | List | Tuple`.

In fact I think there are two kinds of `AsName`s - the ones in imports and except handlers can only take a `Name`, while the ones in with statements can accept any target expression (`BaseAssignTargetExpression`).

```
echo 'with x as a[1]: ...' | python -m libcst.tool print -                                                                                                                                                
Module(
  body=[
    With(
      items=[
        WithItem(
          item=Name(
            value='x',
          ),
          asname=AsName(
            name=Subscript(
              value=Name(
                value='a',
              ),
              slice=[
                SubscriptElement(
                  slice=Index(
                    value=Integer(
                      value='1',
                    ),
                  ),
                ),
              ],
            ),
          ),
        ),
      ],
      body=SimpleStatementSuite(
        body=[
          Expr(
            value=Ellipsis(),
          ),
        ],
      ),
    ),
  ],
)
```

**Comment by jakkdl** [comment]
I just hit this one, and wrote up a long and thorough issue before I tried searching for an existing issue and found this one :upside_down_face: 

I'll just copy-paste the stuff I wrote here:
## problem
Currently all three of [ExceptHandler](https://libcst.readthedocs.io/en/latest/nodes.html#libcst.ExceptHandler), [ImportAlias](https://libcst.readthedocs.io/en/latest/nodes.html#libcst.ImportAlias) and [WithItem](https://libcst.readthedocs.io/en/latest/nodes.html#libcst.WithItem) use [AsName](https://libcst.readthedocs.io/en/latest/nodes.html#libcst.AsName), but that is incorrect. `AsName.name` is typed as `Name|Tuple|List`, but only `WithItem` can produce an `AsName` with a `Tuple` or `List`.
```python
# Both of these are syntax errors
from a import b as (c, d)
try:
    ...
except MyException as [e, f]:
    ...
# but this is valid
with my_cm as (g, h):
    ...
```

See
* https://docs.python.org/3/library/ast.html#ast.ExceptHandler
* https://docs.python.org/3/library/ast.html#ast.alias
* https://docs.python.org/3/library/ast.html#ast.withitem

## my instantiation of the problem
For me it leads to incorrect typing, but it might plausibly also lead to other problems when generating code, fuzzing, etc.

## suggested fix
Introduce a second, `WithAsName` class. I also considered `ExceptItem` and `ImportFrom` directly containing a `Name` - but realized that might not work with specifying whitespace etc around the `as`. If purely treating this as a typing issue, one could maybe make `AsName` generic with a `TypeVar` specifying the possible types.

**Comment by zsol** [comment]
> Introduce a second, `WithAsName` class.

I think this is the more correct approach, but can we make it a backwards compatible change at least for valid code?

**Comment by jakkdl** [comment]
> > Introduce a second, `WithAsName` class.
> 
> I think this is the more correct approach, but can we make it a backwards compatible change at least for valid code?

Yeah this is tricky. I think one could get there by introducing two subclasses of `AsName`, `WithAsName` and ... `AsNameName` (idk wth to name it), and if a visitor defines `visit_AsName` then it gets executed by both (and same with matchers, etc). With only a single new class I'm failing to come up with a backwards-compatible change that improves much.

Or go all the way to three different classes, `ImportAsName`, `WithAsName`, `ExceptAsName`.

`AsName` should probably get deprecated at the same time, but I could see an argument for keeping it around.
